embassy-time/src/queue.rs

```diff
@@ -0,0 +1,617 @@
+//! Generic timer queue implementation
+//!
+//! This module provides a timer queue that works with any executor.
+//!
+//! In terms of performance, this queue will likely be less efficient in comparison to executor-native queues,
+//! like the one provided with e.g. the `embassy-executor` crate.
+//!
+//! # Enabling the queue
+//! - Enable the Cargo feature `generic-queue`. This will automatically instantiate the queue.
+//!
+//! # Initializing the queue
+//! - Call ```unsafe { embassy_time::queue::initialize(); }``` early on in your program, before any of your futures that utilize `embassy-time` are polled.
+//!
+//! # Customizing the queue
+//! - It is possible to customize two aspects of the queue:
+//!   - Queue size:
+//!     By default, the queue can hold up to 128 timer schedules and their corresponding wakers. While it will not crash if more timer schedules are added,
+//!     the performance will degrade, as one of the already added wakers will be awoken, thus making room for the new timer schedule and its waker.
+//!   - The mutex (i.e. the [`RawMutex`](embassy_sync::blocking_mutex::raw::RawMutex) implementation) utilized by the queue:
```

Comment:
This feels way too tricky to push onto the user's responsibility. IMO the time driver should be "critical-section compliant", meaning that it won't call arbitrary code (embassy's code) from a context that "skips" the critical-section protections. 

For example, if the CS impl in use is "lock the RTOS scheduler to avoid the current thread getting switched out, but not disable IRQs" then the time driver shouldn't call the alarm callback from an ISR, only from regular threads.

This would mean we can hardcode `CriticalSectionMutex` here. Which is exactly what `critical-section` is for, libs should be able to blindly rely on it.

Reply:
I don't think we can blindly simplify it and hard-code it to "the CS impl the HAL provides to the `critical-section` crate should be compatible with the Driver impl the same HAL provides to the `embassy-time` crate".

For example:
- I believe that the `critical-section` impl that should be provided by the ESP-IDF HAL should be based on a FreeRTOS mutex (which is equivalent to a STD mutex), even if this mutex cannot be used from an ISR context. This is so because in most of the cases user code is **not** executed in an ISR context. (ISR contexts are reserved for the minority, who has the  "honor" to deal with device drivers, timer queues and executors). So for the regular "user-space" cases, FreeRTOS mutex is the right approach:
  - It does not introduce latencies when handling higher prio tasks or interrupts, as user-space code can be interrupted while holding the mutex
  - It can be locked for arbitrary long intervals
  - Arbitrary user code can be executed in the mutex. Including calls to `libc`, like `malloc`
  - Contrast this with all the restrictions imposed by a "disable-all-interrupts" mutex. I for one am not sure what part of the STD API is safe to call while holding such a mutex!

On the other hand, I also believe that the Driver impl provided by the ESP-IDF HAL **should** be calling back from an ISR. This is because this gives you the lowest latency (we are talking high res timers after all!) *and* we control what is being called - the timer queue and then on top only the executor code which schedules the corresponding task if it was not scheduled already. No user code, no `malloc`s, nothing.

So your suggestion to hard-code the CS impl used in the timer queue puts me between a rock and a hard place:
- I either have to expose a "disable-all-interrupts" CS impl for `critical-section`, with all the consequences for the user
- Or I have to use a less efficient Driver impl, which does not call from an ISR.

Hence why I made it configurable, with some defaults.

Reply:
> and we control what is being called

No, you don't. You're calling the `embassy-time` alarm callback, which can can in turn exec arbitrary code from any crate. For example, if:
  - the end user is using the time driver directly (it's public API), or
  - the end user is using the raw executor, the alarm callback calls the `signal_fn` executor callback which is supplied by the user: https://github.com/embassy-rs/embassy/blob/master/embassy-executor/src/raw/mod.rs#L289

And even if the callback executed only `embassy-time` code, I don't think it's `embassy-time`'s responsibility to publicly guarantee "it's OK to call the callback from a context that violates the `critical-section` guarantees"...

> So your suggestion to hard-code the CS impl used in the timer queue puts me between a rock and a hard place:

Yes, I agree it's quite unfortunate... There's a third option: supply a timer queue impl yourself, in that case you *do* control all code that's being called.

Reply:
> > and we control what is being called
> 
> No, you don't. You're calling the `embassy-time` alarm callback, which can can in turn exec arbitrary code from any crate. For example, if:
> 
> * the end user is using the time driver directly (it's public API), or
> * the end user is using the raw executor, the alarm callback calls the `signal_fn` executor callback which is supplied by the user: https://github.com/embassy-rs/embassy/blob/master/embassy-executor/src/raw/mod.rs#L289
> 
> And even if the callback executed only `embassy-time` code, I don't think it's `embassy-time`'s responsibility to publicly guarantee "it's OK to call the callback from a context that violates the `critical-section` guarantees"...

I think I see where you are going with that. *You are not even talking about the generic queue implementation actually*. You are saying - a `Driver` provider should make sure that users can safely use the critical sections provided by the `critical-section` crate in the user code that the `Driver` calls-back via `set_alarm_callback`, as the Driver API has its uses beyond (a) timer queue. And I now agree to that - the opposite would be very surprising indeed. Can you confirm I'm on the right track?

> Yes, I agree it's quite unfortunate... There's a third option: supply a timer queue impl yourself, in that case you _do_ control all code that's being called.

Except what happens on `waker.wake()` but we can probably assume it is a minimal piece of code that does just task scheduling.
I'll try to expose a minimal public API where the user can take the generic `Queue` impl and instantiate it with their own alarm callbacks that are disconnected from the `Driver` ones. If this proves too unwieldy, oh well, we'll just copy paste.

Reply:
Made the changes and then realized that even with a custom timer queue, `embassy-time` still has a hard-dependency on `Driver` because of the `_embassy_time_now` call, which is provided by the driver, and not by the queue.

Reply:
@Dirbaio Sorry for pinging but this might be the last question remaining with the new design.
Basically:
- Who should provide the `now()` method?
- a) `Driver` as it still does (+) less changes (-) folks contributing their custom timer queue that is not based on `Driver` alarms still need to contribute a `Driver`, just because of the `now()` method provided by the driver
- b) `TimerQueue` (-) a bit more changes (+) somewhat more symmetric (+) `Driver` impl becomes optional, at least w.r.t. contributing a custom `TimerQueue` impl
- c) A third entity - say - a `Now` trait;
- d) Both `TimerQueue` and `Driver` should have a `now` method, but  `Instant::now` will use `TimerQueue::now` instead of `Driver::now`

Reply:
Marking this conversation as resolved. The agreement is to initially go with (a) as it is the simplest variant, and in most cases folks would anyway use either the generic-queue implementation this PR contributes, or the embassy-executor custom one. Both of these require a `Driver` impl anyway.